quickjs.c

```diff
@@ -53310,12 +53322,11 @@ static void reset_weak_ref(JSRuntime *rt, JSWeakRefRecord **first_weak_ref)
             /**
              * During the GC sweep phase the held object might be collected first.
              */
-            if (!JS_IsObject(fre->held_val) || JS_IsLiveObject(frd->ctx->rt, fre->held_val)) {
-                JSValue func = js_dup(frd->cb);
-                JSValue ret = JS_Call(frd->ctx, func, JS_UNDEFINED, 1, &fre->held_val);
-                JS_FreeValueRT(rt, func);
-                JS_FreeValueRT(rt, ret);
-                JS_FreeValueRT(rt, fre->held_val);
+            if (!rt->in_free && (!JS_IsObject(fre->held_val) || JS_IsLiveObject(rt, fre->held_val))) {
+                JSValue args[2];
+                args[0] = frd->cb;
+                args[1] = fre->held_val;
+                JS_EnqueueJob(frd->ctx, js_finrec_job, 2, args);
             }
             JS_FreeValueRT(rt, fre->token);
             js_free_rt(rt, fre);
```

Comment:
This appears to cause #648 and now that I'm looking at it more closely, I believe a spot a bug:

If the branch is _not_ taken (i.e. no job is enqueued), shouldn't it `JS_FreeValueRT(rt, fre->held_val)`? `fre` is freed here.

Reply:
Figured it out just now: it **should** call `JS_FreeValueRT(rt, fre->held_val)` because JS_EnqueueJob increments the ref count, and js_finrec_job should **not** call `JS_FreeValue(ctx, argv[1])`. Pull request incoming.